src/Extensions/Hosting.Services/HostBuilderExtensions.cs

```diff
@@ -0,0 +1,159 @@
+		public static ServiceFabricHostBuilder<TContext> AddServiceListener<TContext>(
+			this ServiceFabricHostBuilder<TContext> builder,
+			string name,
+			Func<TContext, ICommunicationListener> createListener)
+				where TContext : ServiceContext =>
+			builder.AddServiceListener(new ListenerBuilder<TContext>(name, createListener));
+
+
+		/// <summary>
+		/// Add service listener to stateless service fabric service
+		/// </summary>
+		public static ServiceFabricHostBuilder<TContext> AddServiceListener<TContext>(
+			this ServiceFabricHostBuilder<TContext> builder,
+			IListenerBuilder<TContext> listenerBuilder)
+				where TContext : ServiceContext =>
+			builder.ConfigureServices((config, collection) => collection.AddSingleton(listenerBuilder));
+
+
+		/// <summary>
+		/// Configures host to run service fabric stateless service with initializded Omex dependencies
+		/// </summary>
+		public static IHost BuildStatelessService(
+			this IHostBuilder builder,
+			string serviceName,
+			Action<ServiceFabricHostBuilder<StatelessServiceContext>> builderAction) =>
+			builder.BuildServiceFabricService<OmexStatelessServiceRunner, StatelessServiceContext>(serviceName, builderAction);
+
+
+		/// <summary>
+		/// Configures host to run service fabric stateful service with initializded Omex dependencies
+		/// </summary>
+		public static IHost BuildStatefulService(
+			this IHostBuilder builder,
+			string serviceName,
+			Action<ServiceFabricHostBuilder<StatefulServiceContext>> builderAction) =>
+			builder.BuildServiceFabricService<OmexStatefulServiceRunner, StatefulServiceContext>(serviceName, builderAction);
+
+
+		/// <summary>
+		/// Registerin Dependenci Injection classes that will provide Serfice Fabric specific information for logging
+		/// </summary>
+		public static IServiceCollection AddOmexServiceFabricDependencies<TContext>(this IServiceCollection collection)
+			where TContext : ServiceContext
+		{
+			collection.TryAddSingleton<ServiceContextAccessor<TContext>, ServiceContextAccessor<TContext>>();
+			collection.TryAddSingleton<IServiceContextAccessor<TContext>>(p => p.GetService<ServiceContextAccessor<TContext>>());
+			collection.TryAddSingleton<IServiceContextAccessor<ServiceContext>>(p => p.GetService<ServiceContextAccessor<TContext>>());
+			collection.TryAddSingleton<IServiceContext, OmexServiceFabricContext>();
+			collection.TryAddSingleton<IMachineInformation, ServiceFabricMachineInformation>();
+			return collection.AddOmexServices();
+		}
+
+
+		private static IHost BuildServiceFabricService<TRunner,TContext>(
+			this IHostBuilder builder,
+			string serviceName,
+			Action<ServiceFabricHostBuilder<TContext>> builderAction)
+				where TContext : ServiceContext
+				where TRunner : class, IOmexServiceRunner
+		{
+			try
+			{
+				builderAction(new ServiceFabricHostBuilder<TContext>(builder));
+
+				if (string.IsNullOrWhiteSpace(serviceName))
+				{
+					serviceName = Assembly.GetExecutingAssembly().GetName().FullName;
```

Comment:
Do you not need to also call the logic from the `else` branch here too?

Also, `Lazy` seems like it could be more appropriate for this to make the intention clearer

Reply:
Application name could be set in multiple different ways (ex. via an environment variable) and if it won't be set it will have assembly name as the default, so I wanted to preserve this behavior, I've even considered providing build methods without name parameter, but if it's provided it should override default value (logic in else branch).

I'll try to think about a better way of organizing this code and more consistent approach or adding some comments

Reply:
Added more comments to clarify, thanks